Make direct children of a class publicly available

antlechner opened the pull request:
Previously it was only possible to retrieve all children of a class (i.e. those that inherit directly and those that inherit by transitivity) given its ID. The new function retrieves only those children that inherit directly from the given class.

This new function is not used anywhere in CBMC but will be useful for Diffblue code.

I'm not sure how I would write a unit test for this, so if you think a test is necessary, some advice on that would be appreciated. It looks like there are no tests for anything in `class_hierarchy.h` in `cbmc/unit`, but there are some tests related to this file in `jbmc/unit/java_bytecode/goto-programs`. I'm not sure if those tests are supposed to be in the `jbmc` folder as the code in `goto-programs` is not Java-specific? In any case, the change in this PR is quite small so maybe a test is not necessary. Let me know what you think.

antlechner commented:
@peterschrammel @mgudemann @romainbrenguier Could you have a look at this please?

tautschnig commented:
Could unit tests please be added (there are some for this code in jbmc/unit/java_bytecode/goto-programs/class_hierarchy_output.cpp already)?

antlechner commented:
@tautschnig I asked about that in the PR description. :) So you think it's fine to add unit tests for cbmc code in the jbmc unit folder (and the already existing tests are in the right place)?

tautschnig reviewed (commented):
Wouldn't this be rather trivial if using `class_hierarchy_grapht` instead? Then it's just `node.out` ?!

tautschnig commented:
> So you think it's fine to add unit tests for cbmc code in the jbmc unit folder (and the already existing tests are in the right place)?

Then at least it's tested somewhere :-) Yes, I think that would be a perfectly ok place.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: c682411cd13dd988e553441935526d618ee22e03).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77345335

antlechner commented:
@tautschnig I didn't know much about the difference between `class_hierarchyt` and `class_hierarchy_grapht` and why we have both, so I asked @smowton about this in person. If I understand correctly, `class_hierarchyt` is old code and is still used in various places but `class_hierarchy_grapht` is a better representation (as it's graph-based) of the same thing that is generally preferred now.
If that's correct, I could do the following:
1) create helper functions with the same functionality as `class_hierarchyt::get_direct_children` and `class_hierarchyt::get_children_trans` in `class_hierarchy_grapht` (using `node.out` and some conversions between `irep_idt` and `node_indext`)
2) add unit tests for the new functions in `..._grapht`
3) revert the original commit of this PR
4) add documentation to `class_hierarchyt` explaining that it is deprecated and `class_hierarchy_grapht` should be used instead whenever possible, to make sure others like me won't get confused about this in the future.
Does this sound like what you had in mind?

tautschnig commented:
> Does this sound like what you had in mind?

You seem to be going for a grand plan :-) - I'd be totally fine with you using `class_hierarchy_grapht` as-is. My slight pushback on your initial proposal is all based on not growing the code base when (seemingly) not necessary. If, however, you can afford the extra time then surely what you are suggesting is great.

antlechner committed:
Add useful functions to class_hierarchy_grapht

class_hierarchy_grapht is a more advanced representation of the class
hierarchy than class_hierarchyt. So ideally we want to migrate all
functions from the old version to the graph-based representation. This
commit does this for the functions get_parents_trans,
get_children_trans, and introduces a new function get_direct_children.

antlechner committed:
Move non-graph function below graph functions

This way everything related to class_hierarchy_grapht appears first in
the file, followed by everything related to class_hierarchyt.

Ideally these parts should be split into different files in the future.

antlechner committed:
Add documentation to class_hierarchyt

antlechner commented:
@tautschnig That makes sense. From my point of view, of course it's much nicer to have a clean interface of CBMC structures that we can use directly in Diffblue code without having to understand the implementation details of those structures and use five lines of set-up code every time we use them. So I hope my new approach is a good compromise. Of course it adds even more to the code base than my previous approach... But the idea is that it should be a step on the way to move everything in `class_hierarchyt` to `class_hierarchy_grapht`, and once that is all done and the old `class_hierarchyt` is removed, everything should be in a tidier state than before. So hopefully this change will be useful for everyone in the long run. :)
I've done steps 1, 3 and 4 from my plan above. I'll add unit tests for the new functions tomorrow.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 4c650a264bef71cc59c576666b647b65380a9ac2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77381494

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 3cd0c3c5361e4ef9a8ac063bf210c80be5936c90).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77450890

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 11b8026e60577fcbdf0694fdeb438ec4d1259596).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77454640

antlechner committed:
Fix existing class hierarchy test syntax

These two tests were previously never run because of a missing comma in
the SCENARIO line. The two scenarios also need to have different names.

antlechner committed:
Add tests for new class_hierarchy_grapht functions

antlechner commented on src/goto-programs/class_hierarchy.cpp:
```diff
@@ -13,6 +13,7 @@ Date: April 2016
 
 #include "class_hierarchy.h"
 
+#include <iterator>
```
This line was only needed to make the "AWS cbmc-windows" build happy, which would otherwise complain about `back_inserter` in `ids_from_indices`.

antlechner commented:
@tautschnig Step 2 is done - I added unit tests for all the new functions (except helper functions). I also noticed that the previously existing tests for class hierarchies were never actually run so I fixed that. Could you please re-review?

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 6f861fcbbd195818e8326869492d0fac1383c5e5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77479794
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: ff7b37c752dd4145054bb55c8eb061aff908aaa4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77480971
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 9af7ef1356613aacb68ad8791673a32d95006a6d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77481372

tautschnig commented:
@antlechner Thanks a lot for all the work, this looks all very good to me. Just one bit:
> [...] I also noticed that the previously existing tests for class hierarchies were never actually run so I fixed that.

1) How come these were not being run and 2) which change is it that now makes them run?

tautschnig reviewed (approved):
Looks good to me as said, just that question (https://github.com/diffblue/cbmc/pull/2482#issuecomment-400722562) about unit tests remains open.

antlechner commented:
@tautschnig Glad I found a way to make this change that works for everyone!

> which change is it that now makes them run?

The change is in its own commit: https://github.com/diffblue/cbmc/pull/2482/commits/79cad15ba6e1ad3b3ea0f4ef0272e9ab181fdd33
(Or in case that should disappear after a force-push, it's the one called "Fix existing class hierarchy test syntax".)

> How come these were not being run

They used the wrong syntax in the `SCENARIO` line: a comma was missing so the tags weren't applied properly. It's also apparently not ok for two scenarios to have the same name (I tried just adding the commas first and got errors about the names being identical when trying to run the tests).

tautschnig commented:
> The change is in its own commit: 79cad15

Oh, wow, slightly scary that syntax errors lead to silently ignored unit tests.

antlechner commented:
Agreed! I `grep`ed for that particular mistake in both `unit` and `jbmc/unit` and luckily couldn't find any other occurrences of it.
